### PR TITLE
re-factor CI workflows

### PR DESCRIPTION
**Pull request by overcookedpanda**
This PR includes 2 small fixes for MinGW compilation, along with re-working the CI workflows to include a nightly build at midnight UTC with full gui, and full tests being run.  For standard workflow, changes to the develop branch will run without gui and without tests, and changes to the master branch will build the gui, but will not run tests.

### CODE REVIEW

File: depends/packages/altintegration.mk [hunk 1]
```diff
@@ -5,7 +5,7 @@ $(package)_file_name=$($(package)_version).tar.gz
 $(package)_sha256_hash=543128f8483f8a6a4bb78f203e37989151463318d8de81cad3fb08ce1cb45d03
 
 define $(package)_config_cmds
-  cmake -DCMAKE_INSTALL_PREFIX=$($(package)_staging_dir)$(host_prefix) -DCMAKE_BUILD_TYPE=Release -DTESTING=OFF -DWITH_ROCKSDB=OFF -DSHARED=OFF -B .
+  cmake -DCMAKE_C_COMPILER=x86_64-w64-mingw32-gcc -DCMAKE_CXX_COMPILER=x86_64-w64-mingw32-g++ -DCMAKE_INSTALL_PREFIX=$($(package)_staging_dir)$(host_prefix) -DCMAKE_BUILD_TYPE=Release -DTESTING=OFF -DWITH_ROCKSDB=OFF -DSHARED=OFF -B .
```
Comment (Warchant) [inline depends/packages/altintegration.mk]:
```suggestion
  cmake -DCMAKE_C_COMPILER=$(CC) -DCMAKE_CXX_COMPILER=$(CXX) -DCMAKE_INSTALL_PREFIX=$($(package)_staging_dir)$(host_prefix) -DCMAKE_BUILD_TYPE=Release -DTESTING=OFF -DWITH_ROCKSDB=OFF -DSHARED=OFF -B .
```

File: .github/workflows/ci-develop.yml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      - master
+
+jobs:
+  linux:
+    runs-on: ubuntu-latest
+    steps:
+      - name: checkout
+        uses: actions/checkout@v1
+        with:
+          submodules: true
```
Comment (Warchant) [inline .github/workflows/ci-develop.yml]: you can remove submodule

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -3,53 +3,86 @@ name: CI
...
-      run: CONFIG_SITE=$PWD/depends/${{ matrix.host }}/share/config.site ./configure
-        --without-gui
-        --enable-werror
-        --disable-bench
-        #--enable-lcov
-        #--enable-lcov-branch-coverage
+      - name: checkout
+        uses: actions/checkout@v1
+        with:
+          submodules: true
```
Comment (Warchant) [inline .github/workflows/ci.yml]: this can be removed

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -3,53 +3,86 @@ name: CI
...
+      - name: install alt-integration-cpp
+        run: |
+          (
+          cd /opt;
+          wget https://github.com/VeriBlock/alt-integration-cpp/archive/${ALTINTEGRATION_VERSION}.tar.gz;
+          tar -xf ${ALTINTEGRATION_VERSION}.tar.gz;
+          cd alt-integration-cpp-${ALTINTEGRATION_VERSION};
+          mkdir build;
+          cd build;
+          cmake .. -DCMAKE_BUILD_TYPE=Release;
```
Comment (Warchant) [inline .github/workflows/ci.yml]:
```suggestion
          cmake .. -DCMAKE_BUILD_TYPE=Release -DTESTING=OFF -DSHARED=OFF -DWITH_ROCKSDB=OFF;
```

File: .github/workflows/nightly.yml [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+      - name: install alt-integration-cpp
+        run: |
+          (
+          cd /opt;
+          wget https://github.com/VeriBlock/alt-integration-cpp/archive/${ALTINTEGRATION_VERSION}.tar.gz;
+          tar -xf ${ALTINTEGRATION_VERSION}.tar.gz;
+          cd alt-integration-cpp-${ALTINTEGRATION_VERSION};
+          mkdir build;
+          cd build;
+          cmake .. -DCMAKE_BUILD_TYPE=Release -DTESTING=OFF;
```
Comment (Warchant) [inline .github/workflows/nightly.yml]:
```suggestion
          cmake .. -DCMAKE_BUILD_TYPE=Release -DTESTING=OFF -DWITH_ROCKSDB=OFF -DSHARED=OFF;
```

File: Dockerfile [hunk 6]
```diff
@@ -1,9 +1,17 @@
...
-ENV CONFIG_SITE=/app/depends/x86_64-pc-linux-gnu/share/config.site
+RUN export ALTINTEGRATION_VERSION=$(awk -F '=' '/\$\(package\)_version/{print $NF}' $PWD/depends/packages/altintegration.mk | head -n1); \
+    (\
+     cd /opt; \
+     wget https://github.com/VeriBlock/alt-integration-cpp/archive/${ALTINTEGRATION_VERSION}.tar.gz; \
+     tar -xf ${ALTINTEGRATION_VERSION}.tar.gz; \
+     cd alt-integration-cpp-${ALTINTEGRATION_VERSION}; \
+     mkdir build; \
+     cd build; \
+     cmake .. -DCMAKE_BUILD_TYPE=Release -DTESTING=OFF; \
```
Comment (Warchant) [inline Dockerfile]:
```suggestion
     cmake .. -DCMAKE_BUILD_TYPE=Release -DTESTING=OFF -DWITH_ROCKSDB=OFF -DSHARED=OFF; \
```

Comment (overcookedpanda) [reply Dockerfile]: I've fixed this locally, and will push with next fix.

File: depends/packages/altintegration.mk (hunk 1 above)
Comment (overcookedpanda) [reply depends/packages/altintegration.mk]: To be able to fix this, we will have to figure out how to set the cmake compiler options from the mingw32.mk file.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -3,49 +3,112 @@ name: CI
...
-          sudo update-alternatives --set x86_64-w64-mingw32-g++ /usr/bin/x86_64-w64-mingw32-g++-posix
-          sudo update-alternatives --install /usr/bin/x86_64-w64-mingw32-gcc x86_64-w64-mingw32-gcc /usr/bin/gcc 90
-          sudo update-alternatives --install /usr/bin/x86_64-w64-mingw32-g++ x86_64-w64-mingw32-g++ /usr/bin/g++ 90
+          current_date=`date --utc +%Y%m%d_%H%M%SZ`
+          echo "::set-output name=timestamp::${current_date}"
+      - name: ccache cache files
+        uses: actions/cache@v1.1.0
+        with:
+          path: ~/.ccache
+          key: linux-ccache-${{ steps.cache_timestamp.outputs.timestamp }}
```
Comment (Warchant) [inline .github/workflows/ci.yml:31]: do you think timestamp is good idea? 
given 2 builds, their timestamps will be different -> no cache will be reused

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -3,49 +3,112 @@ name: CI
 on:
   push:
     branches:
+      - develop
```
Comment (Warchant) [inline .github/workflows/ci.yml:6]: when you do this, your PR will double number of builds from this file, as you both `push` to develop and opened a PR with base = develop.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -3,49 +3,112 @@ name: CI
...
+          CC: x86_64-w64-mingw32-gcc
+          CXX: x86_64-w64-mingw32-g++
+          HOST: x86_64-w64-mingw32
+          BASE_CACHE: ~/.depends_cache
+        run: cd depends && make -j2 NO_QT=1
       - name: autogen
         run: ./autogen.sh
       - name: configure
-        run: CONFIG_SITE=$PWD/depends/${{ matrix.host }}/share/config.site ./configure
+        run: CONFIG_SITE=$PWD/depends/x86_64-w64-mingw32/share/config.site ./configure --prefix=/
```
Comment (Warchant) [inline .github/workflows/ci.yml:115]:
```suggestion
        run: CONFIG_SITE=$PWD/depends/$HOST/share/config.site ./configure --prefix=/
```

Comment (Warchant) [reply .github/workflows/ci.yml:115]: also, is it **really** necessary to set prefix=/ ?

File: .github/workflows/ci.yml (hunk 7 above)
Comment (overcookedpanda) [reply .github/workflows/ci.yml:31]: Cache is re-used during my tests, so this should be no issue.  It restores based on name, following other examples I included the timestamp, but it has no effect as to it being used..

File: .github/workflows/ci.yml (hunk 9 above)
Comment (overcookedpanda) [reply .github/workflows/ci.yml:115]: I feel that we should stick closely to the BTC core instructions which include setting the --prefix=/.

image

File: .github/workflows/ci.yml (hunk 8 above)
Comment (overcookedpanda) [reply .github/workflows/ci.yml:6]: We can adjust the triggers, I was just using the same as the currently existing triggers on in ci.yml.

**Review by Warchant — APPROVED** [review]

**Pull request merged**